File: src/main/scala/com/nthportal/shell/async/SimpleInputChannel.scala
package com.nthportal.shell.async

import com.nthportal.collection.concurrent.FutureQueue

import scala.concurrent.Future

/**
  * A simple [[InputChannel]] implementation.
  */
final class SimpleInputChannel extends InputChannel {
  private val queue = FutureQueue[InputAction[_]]()

  override def sendAction[T](action: InputAction[T]): Future[T] = {
    queue += action
    action.future
  }

  override def nextAction: Future[InputAction[_]] = queue.dequeue()
}

object SimpleInputChannel {
  def apply(): SimpleInputChannel = new SimpleInputChannel()
}

File: src/main/java/com/nthportal/shell/async/compat/SimpleInputChannel.java
package com.nthportal.shell.async.compat;

import com.nthportal.collection.concurrent.FutureQueue;

import java.util.concurrent.CompletionStage;

import static scala.compat.java8.FutureConverters.toJava;

/**
 * A simple {@link InputChannel} implementation.
 */
public final class SimpleInputChannel implements InputChannel {
    private final FutureQueue<InputAction<?>> queue = FutureQueue.empty();

    @Override
    public <T> CompletionStage<T> sendAction(InputAction<T> action) {
        queue.enqueue(action);
        return toJava(action.future());
    }

    @Override
    public CompletionStage<InputAction<?>> nextAction() {
        return toJava(queue.dequeue());
    }
}
